File: decl-ts/src/qengine/revisions.ts
// Revision verification over the Engine's existing slot cache. Previous values
// live here only while an invalidated slot awaits verification or recomputation.
import type { Engine } from '../engine.ts';
import { isArr, isMap, isQ, isRange, isRec, isRef, pathStr, valueEq } from '../semantics.ts';
import type { RecInst, Value } from '../semantics.ts';

interface Pending {
  value: Value;
  deps: Set<string>;
  changed: number;
  forced: boolean;
  matches?: (value: Value) => boolean;
}

/** Values with no constructed records or frozen owners can be compared without
 * forcing another query or hiding a change of structural ownership. */
function comparable(v: Value): boolean {
  if (!v || typeof v !== 'object') return true;
  if (isRec(v)) return false;
  if (isRef(v)) return !(v.snap || v.inverseRef || v.roundRef);
  if (isArr(v)) return v.items.every(comparable);
  if (isMap(v)) return [...v.entries.values()].every(comparable);
  return isQ(v) || isRange(v);
}

function equal(a: Value, b: Value): boolean {
  if (isArr(a) && isArr(b))
    return (
      pathStr(a.path) === pathStr(b.path) &&
      a.items.length === b.items.length &&
      a.items.every((v: Value, i: number) => equal(v, b.items[i]))
    );
  if (isMap(a) && isMap(b)) {
    const ae = [...a.entries],
      be = [...b.entries];
    return (
      pathStr(a.path) === pathStr(b.path) &&
      ae.length === be.length &&
      ae.every(([k, v], i) => k === be[i][0] && equal(v, be[i][1]))
    );
  }
  return valueEq(a, b);
}

export class Revisions {
  revision = 0;
  verifiedCutoffs = 0;
  valueCutoffs = 0;
  recomputed = 0;
  private readonly changed = new Map<string, number>();
  private readonly pending = new Map<string, Pending>();
  resolve: ((eng: Engine, key: string) => boolean) | null = null;

  get trackedQueries(): number {
    return this.changed.size;
  }

  prune(eng: Engine): void {
    for (const key of this.changed.keys())
      if (!eng.reads.has(key) && !eng.slotsByKey.has(key)) this.changed.delete(key);
  }

  beginQueries(
    eng: Engine,
    invalid: Set<string>,
    values: Map<string, Value>,
    forced: Set<string>,
    capture: (value: Value) => (next: Value) => boolean,
  ): void {
    this.pending.clear();
    this.revision++;
    for (const key of invalid) {
      if (values.has(key)) {
        const value = values.get(key);
        this.pending.set(key, {
          value,
          deps: new Set(eng.reads.get(key)),
          changed: this.changed.get(key) ?? 0,
          forced: forced.has(key),
          matches: capture(value),
        });
      }
      this.changed.set(key, this.revision);
    }
  }

  force(key: string): void {
    const prior = this.pending.get(key);
    if (prior) prior.forced = true;
    this.changed.set(key, this.revision);
  }

  accept(key: string, value: Value): void {
    const prior = this.pending.get(key);
    if (
      prior &&
      (prior.matches ? prior.matches(value) : comparable(value) && equal(prior.value, value))
    ) {
      this.changed.set(key, prior.changed);
      this.valueCutoffs++;
    }
    this.pending.delete(key);
  }

  finish(): void {
    this.pending.clear();
  }

  /** External observations advance first. A recomputed equal slot restores its
   * earlier change stamp, letting its readers verify without running again. */
  begin(eng: Engine, invalid: Set<string>, forced: Set<string>, dropped: Set<RecInst>): void {
    this.pending.clear();
    this.revision++;
    for (const key of invalid) {
      const found = eng.slotsByKey.get(key),
        slot = found?.inst.slots.get(found.name);
      if (
        found &&
        !dropped.has(found.inst) &&
        slot?.state === 'ok' &&
        slot.compute &&
        comparable(slot.value)
      )
        this.pending.set(key, {
          value: slot.value,
          deps: new Set(eng.reads.get(key)),
          changed: this.changed.get(key) ?? 0,
          forced: forced.has(key),
        });
      this.changed.set(key, this.revision);
    }
  }

  /** Called inside the normal slot step, so cycles, deferral and diagnostics
   * retain their normal forcing stack. Verification reads are not new reads of
   * the caller; recomputed dependencies still record their own dependencies. */
  compute(eng: Engine, key: string, run: () => Value): Value {
    const prior = this.pending.get(key);
    if (!prior) return run();
    const unchanged =
      !prior.forced &&
      eng.verifyReads(() => {
        // Verify owners before their former contents. A removed record must
        // not execute merely to verify a reader's old dependency list.
        const deps = [...prior.deps].sort();
        deps.sort((a, b) => Number(b.startsWith('root:')) - Number(a.startsWith('root:')));
        for (const dep of deps) {
          if (this.pending.has(dep)) {
            if (this.resolve) {
              if (!this.resolve(eng, dep)) return false;
            } else {
              const found = eng.slotsByKey.get(dep);
              if (!found) return false;
              eng.forceSlot(found.inst, found.name);
            }
          }
          if ((this.changed.get(dep) ?? 0) === this.revision) return false;
        }
        return true;
      });
    if (unchanged) {
      eng.reads.set(key, prior.deps);
      this.changed.set(key, prior.changed);
      this.pending.delete(key);
      this.verifiedCutoffs++;
      return prior.value;
    }
    const value = run();
    this.recomputed++;
    this.accept(key, value);
    return value;
  }
}

File: decl-py/src/decl/qengine/revisions.py
"""Revision verification over the Engine's existing slot cache.

Previous values are held only while a slot awaits verification/recomputation.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from decl.semantics import ABSENT, ArrV, MapV, Quantity, RangeV, RecInst, Ref, value_eq

if TYPE_CHECKING:
    from decl.engine import Engine


@dataclass
class Pending:
    value: Any
    deps: set[str]
    changed: int
    forced: bool
    matches: Callable[[Any], bool] | None = None


def comparable(v: Any) -> bool:
    if v is ABSENT or v is None or type(v) in (int, float, str, bool):
        return True
    if isinstance(v, RecInst):
        return False
    if isinstance(v, Ref):
        return not (v.snap or v.inverse or v.round_ref)
    if isinstance(v, ArrV):
        return all(comparable(x) for x in v.items)
    if isinstance(v, MapV):
        return all(comparable(x) for x in v.entries.values())
    return isinstance(v, (Quantity, RangeV))


def equal(a: Any, b: Any) -> bool:
    if isinstance(a, ArrV) and isinstance(b, ArrV):
        return (
            a.path == b.path
            and len(a.items) == len(b.items)
            and all(equal(x, y) for x, y in zip(a.items, b.items, strict=True))
        )
    if isinstance(a, MapV) and isinstance(b, MapV):
        return (
            a.path == b.path
            and list(a.entries) == list(b.entries)
            and all(equal(v, b.entries[k]) for k, v in a.entries.items())
        )
    return value_eq(a, b)


class Revisions:
    def __init__(self) -> None:
        self.revision = 0
        self.verified_cutoffs = 0
        self.value_cutoffs = 0
        self.recomputed = 0
        self.changed: dict[str, int] = {}
        self.pending: dict[str, Pending] = {}
        self.resolve: Callable[[Engine, str], bool] | None = None

    @property
    def tracked_queries(self) -> int:
        return len(self.changed)

    def prune(self, eng: Engine) -> None:
        self.changed = {
            key: stamp
            for key, stamp in self.changed.items()
            if key in eng.reads or key in eng.slots_by_key
        }

    def begin_queries(
        self,
        eng: Engine,
        invalid: set[str],
        values: dict[str, Any],
        forced: set[str],
        capture: Callable[[Any], Callable[[Any], bool]],
    ) -> None:
        self.pending.clear()
        self.revision += 1
        for key in invalid:
            if key in values:
                value = values[key]
                self.pending[key] = Pending(
                    value,
                    set(eng.reads.get(key, ())),
                    self.changed.get(key, 0),
                    key in forced,
                    capture(value),
                )
            self.changed[key] = self.revision

    def force(self, key: str) -> None:
        prior = self.pending.get(key)
        if prior is not None:
            prior.forced = True
        self.changed[key] = self.revision

    def accept(self, key: str, value: Any) -> None:
        prior = self.pending.pop(key, None)
        if prior is not None and (
            prior.matches(value)
            if prior.matches
            else comparable(value) and equal(prior.value, value)
        ):
            self.changed[key] = prior.changed
            self.value_cutoffs += 1

    def finish(self) -> None:
        self.pending.clear()

    def begin(
        self, eng: Engine, invalid: set[str], forced: set[str], dropped: set[RecInst]
    ) -> None:
        self.pending.clear()
        self.revision += 1
        for key in invalid:
            found = eng.slots_by_key.get(key)
            slot = found[0].slots.get(found[1]) if found else None
            if (
                found
                and found[0] not in dropped
                and slot
                and slot.state == "ok"
                and slot.compute
                and comparable(slot.value)
            ):
                self.pending[key] = Pending(
                    slot.value,
                    set(eng.reads.get(key, ())),
                    self.changed.get(key, 0),
                    key in forced,
                )
            self.changed[key] = self.revision

    def compute(self, eng: Engine, key: str, run: Callable[[], Any]) -> Any:
        prior = self.pending.get(key)
        if prior is None:
            return run()

        def verify() -> bool:
            # Verify owners before their former contents: deleted records
            # must not execute while verifying a reader's old dependencies.
            for dep in sorted(prior.deps, key=lambda d: (not d.startswith("root:"), d)):
                if dep in self.pending:
                    if self.resolve:
                        if not self.resolve(eng, dep):
                            return False
                    else:
                        found = eng.slots_by_key.get(dep)
                        if found is None:
                            return False
                        eng.force_slot(*found)
                if self.changed.get(dep, 0) == self.revision:
                    return False
            return True

        if not prior.forced and eng.verify_reads(verify):
            eng.reads[key] = prior.deps
            self.changed[key] = prior.changed
            del self.pending[key]
            self.verified_cutoffs += 1
            return prior.value
        value = run()
        self.recomputed += 1
        self.accept(key, value)
        return value
